libraries/persistent_store/src/store.rs

```diff
@@ -128,3 +168,1089 @@ impl StoreUpdate {
+    /// - The entry starts at the beginning of a virtual page.
+    /// - The entry spans 2 pages.
+    fn init_page(&mut self, first: Position, last: Position) -> StoreResult<()> {
+        debug_assert!(first <= last);
+        debug_assert!(last - first < self.format.virt_page_size());
+        let new_first = if first.word(&self.format) == 0 {
+            first
+        } else if first.page(&self.format) == last.page(&self.format) {
+            return Ok(());
+        } else {
+            last + 1
+        };
+        let page = new_first.page(&self.format);
+        if let WordState::Valid(_) = self.parse_init(page)? {
+            return Ok(());
+        }
+        let index = self.format.index_init(page);
+        let init_info = self.format.build_init(InitInfo {
+            cycle: new_first.cycle(&self.format),
+            prefix: new_first.word(&self.format),
+        });
+        self.storage_write_slice(index, &init_info)?;
+        Ok(())
+    }
+
+    /// Sets the padding bit of a user header.
+    fn set_padding(&mut self, pos: Position) -> StoreResult<()> {
+        let mut word = Word::from_slice(self.read_word(pos));
+        self.format.set_padding(&mut word);
+        self.write_slice(pos, &word.as_slice())?;
+        Ok(())
+    }
+
+    /// Sets the deleted bit of a user header.
+    fn set_deleted(&mut self, pos: Position) -> StoreResult<()> {
+        let mut word = Word::from_slice(self.read_word(pos));
+        self.format.set_deleted(&mut word);
+        self.write_slice(pos, &word.as_slice())?;
+        Ok(())
+    }
+
+    /// Wipes a slice of words.
+    fn wipe_span(&mut self, pos: Position, length: Nat) -> StoreResult<()> {
+        let length = (length * self.format.word_size()) as usize;
+        self.write_slice(pos, &vec![0x00; length])
+    }
+
+    /// Returns an extremum page.
+    ///
+    /// With `Greater` returns the most recent page (or the tail). With `Less` returns the oldest
+    /// page (or the head).
+    fn get_extremum_page_head(&self, ordering: Ordering) -> StoreResult<Position> {
+        let mut best = None;
+        for page in 0..self.format.num_pages() {
+            let init = match self.parse_init(page)? {
+                WordState::Valid(x) => x,
+                _ => continue,
+            };
+            let pos = self.format.page_head(init, page);
+            if best.map_or(true, |x| pos.cmp(&x) == ordering) {
+                best = Some(pos);
+            }
+        }
+        // There is always at least one initialized page.
+        best.ok_or(StoreError::InvalidStorage)
+    }
+
+    /// Returns the number of words that can be written without compaction.
+    ///
+    /// This can be temporarily negative during compaction.
+    fn immediate_capacity(&self) -> StoreResult<isize> {
+        let tail = self.tail()?;
+        let end = self.head()? + self.format.virt_size();
+        Ok(if tail > end {
+            -((tail - end) as isize)
```

Comment:
This seems to be the only reason to use `isize`. No callers use the negative value, so you could also return 0 in this case. Would that make sense from what you expect from a "capacity"?

Reply:
Good point, done.

The only problem I would see with returning 0 is that one cannot call `prepare` to finish a compaction cycle on a full store (very rare and probably meaningless circumstance).